### ISSUE TITLE
backward pass of na3d() is very slow

### ISSUE THREAD

**Issue by nikhilmishra000** [SHI-Labs/NATTEN#157]
The backward pass of `na3d()` seems very slow even for modest problem sizes:

Minimal benchmark code:
```
import time
import natten
from natten.functional import *

natten.use_fused_na(True)

b = 1
n_heads, head_dim = 4, 64
t, h, w = 32, 32, 32

for i in range(5):
    print(f"----------\ntrial {i}")
    query = torch.randn(b, t, h, w, n_heads, head_dim, device='cuda', requires_grad=True)
    keys, values = torch.randn_like(query), torch.randn_like(query)

    torch.cuda.synchronize()
    start = time.time()
    out = na3d(query, keys, values, kernel_size=3)
    end = time.time()
    print(f"forward: {end - start:3f}")

    loss = out.sum()

    torch.cuda.synchronize()
    start = time.time()
    loss.backward()
    torch.cuda.synchronize()
    end = time.time()
    print(f"backward: {end - start:3f}")
```

Which results in:
```
----------
trial 0
forward: 0.027356
backward: 0.494834
----------
trial 1
forward: 0.000348
backward: 0.458075
----------
trial 2
forward: 0.000355
backward: 0.458174
----------
trial 3
forward: 0.000371
backward: 0.458034
----------
trial 4
forward: 0.000343
backward: 0.458047
```

These times are on A100-80GB and I experience similar behavior on H100.
This is all with torch 2.2.0, cuda 12.1, python 3.11.

**Comment by alihassanijr** [comment]
Thank you for your interest.

So to provide a bit of context, the backward pass by default only parallelizes along batch and heads, which for your case means there's only 4 CTAs doing all the work, and on both A100 and H100 that's less than 4% of the total number of SMs.

The reason for that is that by default, NATTEN disables KV parallelism, because it does not guarantee determinism. However, this is very standard for most reduction operations.

Can you try enabling KV parallelism and check again?

```
import natten

natten.use_kv_parallelism_in_fused_na(True)
```

Reference: https://github.com/SHI-Labs/NATTEN/blob/main/docs/fna/kv-parallelism.md


There's also auto-tuning, and that can help quite a bit, because it really matters what your specific 3D tile shape is. Unfortunately, with more dimensions (i.e. 2D over 1D, and 3D over 2D) comes more knobs and more possible kernel configurations, and because there aren't too many good heuristics, NATTEN provides an auto-tuner that can search all available tile shapes, and pick the optimal one.
But auto-tuning, which happens only the first time you give `na3d` a new problem size or dtype, can take a few seconds and up to a few minutes, depending on the settings.

I'll leave a link to that here, but we can revisit that once we get past KV parallelism.

**Comment by nikhilmishra000** [comment]
Gotcha, thanks for the explanation! 

Enabling KV parallelism gives me this, which seems reasonable:
```
----------
trial 0
forward: 0.000368
backward: 0.012068
----------
trial 1
forward: 0.000239
backward: 0.012024
----------
trial 2
forward: 0.000205
backward: 0.012047
----------
trial 3
forward: 0.000226
backward: 0.011787
----------
trial 4
forward: 0.000172
backward: 0.012112
```

I tried also enabling the autotuner:
```
natten.use_autotuner(backward_pass=True)
```
but it seems to have not converged after > 10 min. 

I think I'm unblocked with the speed reported above so will prob forgo the autotuning for now. Thanks for the help & the quick response!

**Comment by alihassanijr** [comment]
Oh yeah auto-tuning is just really difficult, and I hope to have a better solution for it. You can technically auto-tune more quickly with fewer iterations and search fewer settings, but yeah it's definitely a feature that needs more work.

```
natten.use_autotuner(
  forward_pass=True,
  backward_pass=True,
  thorough_mode_backward=False,
  warmup_steps_forward=5,
  warmup_steps_backward=5,
  steps_forward=5,
  steps_backward=5,
)
```

If you're interested, there's also a natten environment variable that you can set so autotuner spits out its progress:
```
export NATTEN_LOG_LEVEL="debug"
```

But anyway, I'm very glad your issue was resolved.

**Comment by universome** [comment]
I've encountered the same problem (I also use temporally causal attention):
- by default, forward+backward with na3d is ~5000 times (sic!) slower, than just the forward (but should be ~3 times slower)
- using `natten.use_kv_parallelism_in_fused_na(True)`, made forward+backward "just" ~10 times slower
Unfortunately, that speed is still unaffordable for my setup...

**Comment by alihassanijr** [comment]
Could you please share your use case (problem size, dtype, GPU)?

**Comment by alihassanijr** [comment]
Sorry by channel size 4 you mean dim per attention head is 4?

**Comment by universome** [comment]
sorry for the delay,
basically, i was experimenting with just a single head of dim=4 and single head of dim=8 (for bf16)
i later found that if i move things into low-resolution high-dim regime, then forward+backward takes "just" 3.7 times slower than forward-only — still not ideal, but much better than before
right now, i'm experimenting with other attention variants (i.e. not neighbourhood ones)
